### PR TITLE
Update the docs for Allflex Product Mapper

### PR DESCRIPTION
**Pull request by jacobtani**

### CODE REVIEW

**Comment by jacobtani** [comment]
[(0.5) Document the Allflex mapper](https://trello.com/c/jaXfbxuf/1850-05-document-the-allflex-mapper)

File: README.md [hunk 1]
```diff
@@ -43,19 +44,25 @@ product.success?  # => true
...
+There are three kinds of formatter dependencies you can optionally inject:
+  * The `NilFormatter` does what it says.  
+  * The `XmlFormatter` returns Allflex XML.  Not the complete order, mind, rather the XML for an individual product.  For marked tag that will be a `<MarkedOrderLine>`, for an accessory, that will be a `<BlankOrderLine>`.  Check out `./spec/shared/examples/allflex_xml.rb` for examples. You will see the various XML components being generated here using the data computed eg `OrderLine`
+  * `XmlTypeSplittingFormatter`: Returns the Allflex XML and also indicates whether it is a `MarkedOrderLine` or `BlankOrderLine` product to help the correct element around each OrderLine from the external service that consumes the mapper.
 
+```ruby
+product = Allflex::ProductMapper::Mapper.new.(ordered_product ::Allflex::ProductMapper::XmlTypeSplittingFormatter)
```
Comment (josephleniston) [inline README.md]:
```suggestion
result = ::Allflex::ProductMapper::Mapper.new.(ordered_product, ::Allflex::ProductMapper::XmlTypeSplittingFormatter)
```

File: README.md [hunk 2]
```diff
@@ -43,19 +44,25 @@ product.success?  # => true
 product.value_or  # Allflex::ProductMapper::Product
 ```
 
-You can also optionally pass in a formatter.  The `NilFormatter` does what it says.  The `XmlFormatter` returns Allflex XML.  Not the complete order, mind, rather the XML for an individual product.  For marked tag that will be a `<MarkedOrderLine>`, for an accessory, that will be a `<BlankOrderLine>`.  Check out `./spec/shared/examples/allflex_xml.rb` for examples.
+There are three kinds of formatter dependencies you can optionally inject:
+  * The `NilFormatter` does what it says.  
+  * The `XmlFormatter` returns Allflex XML.  Not the complete order, mind, rather the XML for an individual product.  For marked tag that will be a `<MarkedOrderLine>`, for an accessory, that will be a `<BlankOrderLine>`.  Check out `./spec/shared/examples/allflex_xml.rb` for examples. You will see the various XML components being generated here using the data computed eg `OrderLine`
+  * `XmlTypeSplittingFormatter`: Returns the Allflex XML and also indicates whether it is a `MarkedOrderLine` or `BlankOrderLine` product to help the correct element around each OrderLine from the external service that consumes the mapper.
```
Comment (josephleniston) [inline README.md]: > to help the correct element around each OrderLine

Not sure what this means. Can you reword to make it clearer?

File: README.md [hunk 3]
```diff
@@ -43,19 +44,25 @@ product.success?  # => true
 product.value_or  # Allflex::ProductMapper::Product
 ```
 
-You can also optionally pass in a formatter.  The `NilFormatter` does what it says.  The `XmlFormatter` returns Allflex XML.  Not the complete order, mind, rather the XML for an individual product.  For marked tag that will be a `<MarkedOrderLine>`, for an accessory, that will be a `<BlankOrderLine>`.  Check out `./spec/shared/examples/allflex_xml.rb` for examples.
+There are three kinds of formatter dependencies you can optionally inject:
+  * The `NilFormatter` does what it says.
```
Comment (josephleniston) [inline README.md]: Not very friendly. Maybe explain why you would pass this in?

File: README.md [hunk 4]
```diff
@@ -43,19 +44,25 @@ product.success?  # => true
...
 ## Configuring Allflex Products
 
-The Mapper is NOT configuration-free.  There is no algorithmic method to map from an LIC definition of an Allflex product, and the Allflex product defined in the Allflex systems.
+The Mapper is NOT configuration-free.  There is no algorithmic method to map from an LIC definition of an Allflex product, and the Allflex product defined in the Allflex systems.  Mores the pity!
 
 Therefore, the mapping occurs in the following places:
 
-+ `lib/allflex/product_mapper/manufacturing_definition.rb`: Provides the mapping definition for each Allflex product.
-+ `lib/allflex/product_mapper/allflex_product.rb`: Value object that also performs the weird Tag/Pack ID generation.
++ `lib/allflex/product_mapper/manufacturing_definition.rb`: Defines the mappings that will translate the OrderedProduct into an Allflex-mappable form.  The manufacturing definition is mostly applied by `lib/allflex/product_mapper/product_builder.rb`
```
Comment (josephleniston) [inline README.md]: Make these relative links e.g. [manufacturing_definition.rb](lib/allflex/product_mapper/manufacturing_definition.rb)

File: README.md [hunk 5]
```diff
@@ -43,19 +44,25 @@ product.success?  # => true
...
++ `lib/allflex/product_mapper/allflex_product.rb`: This is essentially the Allflex Product value object.  The `product_builder` injects all the mappings into this object, which is then used by any formatter provided.  It also has the responsibility for generating the Allflex product code, which is derived from the product configuration.  So, in this class, you'll also find configuration necessary to construct the product code.  The defined codes are from Allflex API documentation at: `https://allflex.co.nz:8443/api/`
```
Comment (josephleniston) [inline README.md]: Maybe break this up into sections? Mapping, Tag Example, Accessor Example, Mapping Process In Detail.

File: README.md [hunk 6]
```diff
@@ -43,19 +44,25 @@ product.success?  # => true
...
 
-For example: A Premium Calf Pack is a single product in LIC, with 2 size variants (medium and large).  It maps to a single manufacturing definition.  This is always the case; the FulfilmentProduct to ManufacturingDefinition is 1..1
+A HDX Premium Calf Pack is a single marked product in LIC, with 2 size variants (medium and large). It maps to a single manufacturing definition. This is always the case; the FulfilmentProduct to ManufacturingDefinition is 1..1. For this kind of definition the id field is insignificant--because we use the `AllflexProduct` id builder in this case.
```
Comment (josephleniston) [inline README.md]:
```suggestion
A HDX Premium Calf Pack is a single marked product in LIC that comes in two size variants (medium and large).

The pack always maps to a single manufacturing definition. A `FulfilmentProduct` to `ManufacturingDefinition` is `1..1`. For this kind of definition the id field is insignificant. We use the `AllflexProduct` id builder in this case.
```

File: README.md [hunk 7]
```diff
@@ -65,21 +72,57 @@ For example: A Premium Calf Pack is a single product in LIC, with 2 size variant
...
+{
+  id: "00194",
+  product_slug: "allflex-tag-pen"
+}
+```
+
+The general process works like this:
+1) An instance of `Mapper` class is created
+2) A new instance of the passed in formatter is created e.g. `XmlFormatter` and passes in data that is manipulated using the input ordered product.
+3) `OrderedProduct` wraps the LIC product JSON form.  This object provides all methods necessary to extract components from the LIC product form
```
Comment (josephleniston) [inline README.md]: Use either `OrderedProduct` or ordered product consistently.

Also add this clarification (maybe with a link to a fixture example?:
the LIC Ordered Product JSON format known as LIC product form

File: README.md [hunk 8]
```diff
@@ -65,21 +72,57 @@ For example: A Premium Calf Pack is a single product in LIC, with 2 size variant
...
 
+```ruby
+{
+  id: "00194",
+  product_slug: "allflex-tag-pen"
+}
+```
+
+The general process works like this:
+1) An instance of `Mapper` class is created
```
Comment (josephleniston) [inline README.md]: These can also be relative links to classes e.g. [Mapper](lib/allflex/product_mapper/mapper.rb)

File: README.md [hunk 9]
```diff
@@ -65,21 +72,57 @@ For example: A Premium Calf Pack is a single product in LIC, with 2 size variant
...
+2) A new instance of the passed in formatter is created e.g. `XmlFormatter` and passes in data that is manipulated using the input ordered product.
+3) `OrderedProduct` wraps the LIC product JSON form.  This object provides all methods necessary to extract components from the LIC product form
+3) Using this ordered product, the `Fabricator` is called. Its responsibility is to inject the manufacturing definition into the pipeline.  It uses the ordered product's slug to find a product definition with a matching product_slug in `ManufacturingDefinition`. If a matched is found this product definition gets returned.
```
Comment (josephleniston) [inline README.md]:
```suggestion
3) Using this ordered product, the `Fabricator` is called. It's responsibility is to inject the manufacturing definition into the pipeline.  It uses the ordered product's slug to find a product definition with a matching product_slug in `ManufacturingDefinition`. If a match is found this product definition gets returned.
```

File: README.md [hunk 10]
```diff
@@ -65,21 +72,57 @@ For example: A Premium Calf Pack is a single product in LIC, with 2 size variant
...
+4) The `Fabricator` calls `ProductBuilder` is then called with the ordered product and matched manufacturing definition. The bulk of the work happens in this class. First it builds an instance of `AllflexProduct` using the input data. Methods from the `OrderedProduct` class is called to help achieve this.
```
Comment (josephleniston) [inline README.md]:
```suggestion
4) The `Fabricator` then calls `ProductBuilder` with the ordered product and matched manufacturing definition. The bulk of the work happens in this class. First it builds an instance of `AllflexProduct` using the input data. Methods from the `OrderedProduct` class are called to achieve this.
```

File: README.md [hunk 11]
```diff
@@ -31,32 +32,49 @@ bundle update dry-monads
...
+```ruby
+result = Allflex::ProductMapper::Mapper.new.(ordered_product)
+```
 
-product.success?  # => true
+Example of calling the mapper with a formatter option
 
-product.value_or  # Allflex::ProductMapper::Product
+```ruby
+result = ::Allflex::ProductMapper::Mapper.new.(ordered_product,::Allflex::ProductMapper::XmlTypeSplittingFormatter)
```
Comment (josephleniston) [inline README.md]:
```suggestion
result = ::Allflex::ProductMapper::Mapper.new.(ordered_product, ::Allflex::ProductMapper::XmlTypeSplittingFormatter)
```

File: README.md [hunk 12]
```diff
@@ -65,21 +83,60 @@ For example: A Premium Calf Pack is a single product in LIC, with 2 size variant
...
+5) The `AllflexProduct` builds the components contained inside the product using the product slugs in the `marked_tag_mappings` contained in the matching product definition -> the components then call the various builders -> `MarkingsBuilder`, `ColourBuilder`(it matches the URN of the colour) and fetches the size from the input ordered product data.
+7) Each of these builders have an associated marking class they use to process the configuration data i.e. (`MgmtNumberMarking`, `CustomTextMarking`, `BirthIdSeqMarking`, `BirthIdPrefixMarking`, `BlankMarking`, `EidDoNotRemoveLogo`, `PrintTypeMarking`). These markings classes are all an instance of a Dry::Struct class and declare attributes for the marking as well as the name_token which appears in `AllflexProduct` definitions.
+
+As this is not a straightforward process, sometimes the data in the input ordered product JSON is not sufficient to meet the Allflex schema requirements so we have had to implement a number of hacks into this mapper. Namely the following:
```
Comment (josephleniston) [inline README.md]: Make this a seperate section maybe "Mapper hacks"

**Pull request merged**